core/trino-main/src/main/java/io/trino/cost/TopNRankingStatsRule.java

```diff
@@ -0,0 +1,243 @@
+                averageRowsPerPartition,
+                sourceStats,
+                estimateBuilder);
+
+        double outputRowsCount = partitionCount * topRowsPerPartition;
+        return Optional.of(estimateBuilder
+                .setOutputRowCount(outputRowsCount)
+                .addSymbolStatistics(node.getRankingSymbol(), SymbolStatsEstimate.builder()
+                        .setLowValue(1)
+                        // maxRankingPerPartition is the maximum rank per partition. Therefore, in the worse case
+                        // scenario the rank symbol can have maxRankingPerPartition as the high value.
+                        .setHighValue(node.getMaxRankingPerPartition())
+                        .setDistinctValuesCount(rankDistinctCount)
+                        .setNullsFraction(0.0)
+                        .setAverageRowSize(BIGINT.getFixedSize())
+                        .build())
+                .build());
+    }
+
+    private static void fixOrderBySymbolDistinctCount(
+            List<Symbol> orderBy,
+            double partitionCount,
+            double topRowsPerPartition,
+            double averageRowsPerPartition,
+            PlanNodeStatsEstimate sourceStats,
+            PlanNodeStatsEstimate.Builder estimateBuilder)
+    {
+        verify(!orderBy.isEmpty(), "Order by symbols should not be empty for TopNRankingNode.");
+        Symbol firstSortSymbol = orderBy.getFirst();
+        SymbolStatsEstimate symbolStats = sourceStats.getSymbolStatistics(firstSortSymbol);
+        // We assume that the order by columns are uncorrelated to partition columns, hence order by distinct rows
+        // should be equally spread among partitions.
+        double symbolDistinctCount = min(symbolStats.getDistinctValuesCount(), averageRowsPerPartition);
+        // Consider that order by columns are COL1, COL2 for the below single partition:
+        // COL1   COL2
+        // A      1
+        // A      2
+        // A      3
+        // A      4
+        // B      5
+        // B      6
+        // B      7
+        // B      8
+        // The max rows per partition is 8, and for rank 4, the top rows per partition is 4.
+        // So, in that case, the new distinct count for COL1 should be 1. We can calculate the new distinct count
+        // per partition using the formula:
+        // (symbolStats.getDistinctValuesCount() / averageRowsPerPartition) * topRowsPerPartition => (2 / 8) * 4 = 1
+        // Additionally, it's quite hard to argue about for COL2, thus we will not change the distinct count for COL2.
+        // It'll automatically be handled by the StatsNormalizer based on outputRowCount.
+        double newDistinctCountPerPartition = ceil((symbolDistinctCount / averageRowsPerPartition) * topRowsPerPartition);
+        if (isNaN(newDistinctCountPerPartition)) {
+            return;
+        }
+        // Total distinct count should be at most newDistinctCountPerPartition * partitionCount.
```

Comment:
As discussed we most likely shouldn't multiply as previously we assumed order by distinct rows are spread among existing partitions. Since they are also ordered we are hitting same distinct rows in each partition and multiplication is not needed. Good catch. Could you add a comment and remove multiplication